### ISSUE TITLE
Suggestions for improvements

### ISSUE THREAD

**Issue by michaelklachko** [fordDeepDSP/deepSpeech#16]
Hi Lakshmi, 

Thank you for publishing your work! I've been using playing with your code for about a week, and I'd like to discuss ideas for further improvements. As it is, the model reaches about 17% CER (42% WER) so in order to improve it, I'm considering the following:

1. Training on train-clean-360 as well as train-other-500, in addition to train-clean-100. I tried it, and in order to fit processing in memory, I had to compute feats in fp16 precision. So currently I'm running it on clean-100 in fp32 and clean-360 and other-500 in fp16. I trained it with -no-shuffle option, and the model diverges after few hundred steps. I'm not sure yet if this is caused by the reduced precision.

2. Adding a language model. Not sure how to do this. I'm new to speech recognition, and any hints or examples of how this could be added are appreciated!

3.  Instead of a language model, using RNN transducer (a second RNN network to predict output based on previous outputs). This is described in Graves papers, and suggested in DeepSpeech3 paper. Not sure how to implement it: does it need a language model, or does it work on whatever the first RNN outputs (which might not be valid words)?

4. Using LSTM or GRU cells instead of basic RNN. This seems to be straightforward, have you tried it?

5. Using other input processing methods (linear spectrograms, something else?). DeepSpeech2 uses linear spectrograms, but I don't quite understand why it would be better than mfcc. Also, is there any intuition behind choosing particular mfcc parameters (e.g. winlen, numcep, etc)? 

6. Using batch-norm or layer norm. It seems like batch-norm in yao-matrix code is not implemented correctly - it does not even use is_training parameter which should be set to False during testing. I'm not even sure how the batch norm should be implemented for RNN layers. 

What do you think?

**Comment by lscaria** [comment]
Hi @michaelklachko 

Glad to hear that you have found our codebase useful. Thank you for your feedback, let me try to address your points below.

1. Training on more data should help improve the model. However, I am not sure why the model diverges. I wonder if having mixed precision between the clean-100 and the other datasets could be the issue. We will try to look more closely at the effects of using fp16 and get back to you on that.

2. Adding a language model does help improve the CER and WER. I will try to publish a branch which shows how you can use a language model as a simple spell checker/pick the sentence with the highest perplexity after the decoding step, along with instructions on how to create a language model. You can look into [https://github.com/kpu/kenlm](KenLM)

3. While we have not investigated an RNN transducer, We did look at creating our beam search decoders than incorporate language models when decoding. I can publish this in the language model branch so you can see how to incorporate your own implementation of a decoder or an RNN transducer. Decoding is tricky business so if you implement a great decoder please let us know!

4. Using an LSTM layer instead of the RNN does improve results. However it also means many more parameters and slower training. It is pretty simple to modify the RNN cell to use an LSTM instead, but let me know if you run into issues doing so.

5. We have not looked too much into linear spectrograms, but it would definitely be an interesting investigation. As for the mfcc parameters, we kept the default parameters of the function. In some other work with keyword spotting I found those to be the most optimal .

6. Adding batch-norm has been on the to-do list for a while since it was mentioned in DeepSpeech2, however I haven't quite gotten around to it yet.

I hope this answers your questions. We mostly have this repo to support research so that users can easily download, setup, and add their own modifications to the deepSpeech architecture. I will post here once I push a with-decoder branch with the LM and custom decoder, probably in a week or so. I'll also add the LSTM to the master branch.

Best,
Lisa

**Comment by michaelklachko** [comment]
Thanks for the response! I just looked at https://github.com/zzw922cn/Automatic_Speech_Recognition code, which seemed like it could provide some examples for improvement, however it turns out it has a serious issue: they don't use input bucketing, so the model has to take in samples padded to the length of the longest sample in the entire dataset. Which makes it extremely slow. Their batchnorm also seems like they simply put regular batch norm layer after each recurrent layer - I'm not sure how is this supposed to work: https://github.com/zzw922cn/Automatic_Speech_Recognition/blob/master/speechvalley/models/deepSpeech2.py#L59

I am interested in your decoder implementation. Right now I'm looking at Mozilla implementation, where  they create a custom tf op based on a precompiled kenlm module: https://github.com/mozilla/DeepSpeech/blob/31c01db541e6c9d9e8cb86fa766e643ece95c29b/DeepSpeech.py#L488 , but I would prefer importing it in python like this: https://github.com/kpu/kenlm/blob/master/python/example.py because Mozilla's decoder is hard to understand without going through a lot of C++ code...

ASR code has a spellchecker: https://github.com/zzw922cn/Automatic_Speech_Recognition/blob/master/speechvalley/lm/spellingChecker4CN/gardener.py but it's not clear to me how it is integrated.

**Comment by lscaria** [comment]
I added a with_decoder branch along with some documentation in the README there for how we have added an lm and decoder to this deepSpeech implementation. The lm implementation is derived from the kenlm code you linked to above. Please take a look and let us know if there are any issues. Thanks!

**Comment by michaelklachko** [comment]
Thank you, that is very helpful!
